src/controllers/plugins/api/HTTP.cpp

```diff
@@ -0,0 +1,462 @@
+#ifdef CHATTERINO_HAVE_PLUGINS
+#    include "controllers/plugins/api/HTTP.hpp"
+
+#    include "Application.hpp"
+#    include "common/network/NetworkCommon.hpp"
+#    include "common/network/NetworkRequest.hpp"
+#    include "common/network/NetworkResult.hpp"
+#    include "controllers/plugins/LuaAPI.hpp"
+#    include "controllers/plugins/LuaUtilities.hpp"
+#    include "util/DebugCount.hpp"
+
+extern "C" {
+#    include <lauxlib.h>
+#    include <lua.h>
+}
+#    include <QRandomGenerator>
+#    include <QUrl>
+
+#    include <utility>
+
+namespace chatterino::lua::api {
+// NOLINTBEGIN(*vararg)
+// NOLINTNEXTLINE(*-avoid-c-arrays)
+static const luaL_Reg HTTP_REQUEST_METHODS[] = {
+    {"on_success", &HTTPRequest::on_success_wrap},
+    {"on_error", &HTTPRequest::on_error_wrap},
+    {"finally", &HTTPRequest::finally_wrap},
+
+    {"execute", &HTTPRequest::execute_wrap},
+    {"set_timeout", &HTTPRequest::set_timeout_wrap},
+    {"set_payload", &HTTPRequest::set_payload_wrap},
+    {"set_header", &HTTPRequest::set_header_wrap},
+    // static
+    {"create", &HTTPRequest::create},
+    {nullptr, nullptr},
+};
+
+std::shared_ptr<HTTPRequest> HTTPRequest::getOrError(lua_State *L,
+                                                     StackIdx where)
+{
+    if (lua_gettop(L) < 1)
+    {
+        luaL_error(L, "Called c2.HTTPRequest method without a request object");
+        return nullptr;
+    }
+    if (lua_isuserdata(L, where) == 0)
+    {
+        luaL_error(
+            L,
+            "Called c2.HTTPRequest method with a non-userdata 'self' argument");
+        return nullptr;
+    }
+    // luaL_checkudata is no-return if check fails
+    auto *checked = luaL_checkudata(L, where, "c2.HTTPRequest");
+    auto *data =
+        SharedPtrUserData<UserData::Type::HTTPRequest, HTTPRequest>::from(
+            checked);
+    if (data == nullptr)
+    {
+        luaL_error(L, "Called c2.HTTPRequest method with an invalid pointer");
+        return nullptr;
+    }
+    lua_remove(L, where);
+    if (data->target->done)
+    {
+        luaL_error(L, "This c2.HTTPRequest has already been executed!");
+    }
+    return data->target;
+}
+
+void HTTPRequest::createMetatable(lua_State *L)
+{
+    lua::StackGuard guard(L, 1);
+
+    luaL_newmetatable(L, "c2.HTTPRequest");
+    lua_pushstring(L, "__index");
+    lua_pushvalue(L, -2);  // clone metatable
+    lua_settable(L, -3);   // metatable.__index = metatable
+
+    // Generic ISharedResource stuff
+    lua_pushstring(L, "__gc");
+    lua_pushcfunction(L, (&SharedPtrUserData<UserData::Type::HTTPRequest,
+                                             HTTPRequest>::destroy));
+    lua_settable(L, -3);  // metatable.__gc = SharedPtrUserData<...>::destroy
+
+    luaL_setfuncs(L, HTTP_REQUEST_METHODS, 0);
+}
+
+int HTTPRequest::on_success_wrap(lua_State *L)
+{
+    lua::StackGuard guard(L, -2);
+    auto ptr = HTTPRequest::getOrError(L, 1);
+    return ptr->on_success(L);
```

Comment:
`ptr` can be `nullptr` here, but we don't check for it

Reply:
`ptr` here can only be null if we sent Lua a `std::shared_ptr<HTTPRequest>` with a null value. Doing that is not sane. I don't think a check here is necessary.

Reply:
I added one in getOrError anyway with a particularly scary error message.